projects/plugins/jetpack/_inc/client/main.jsx

```diff
@@ -198,6 +199,75 @@ class Main extends React.Component {
 	}
 
 	renderMainContent = route => {
+		/**
+		 * There are two conditions (groups of conditions, really) where we would want to
+		 * show the partner connection screen.
+		 *
+		 * 1. The site is not yet connected to WPCOM, but has the jetpack_partner_coupon
+		 * option set in the database (this.props.partnerCoupon in redux). This is likely a
+		 * partner-user who has just arrived here from a CTA within a partner's dashboard
+		 * or other ecosystem.
+		 *
+		 * 2. The site is already connected to WPCOM, but the jetpack_partner_coupon option
+		 * is still set in the database. This means the user connected their site, but never
+		 * redeemed the coupon. If this is the case, we don't want to override the dashboard
+		 * or at a glance pages with the redemption screen. Instead, we'll catch a URL
+		 * parameter that JITMs will set (showCouponRedemption=true), and show the screen only
+		 * when the user came from a a JITM.
+		 */
+		if ( this.props.partnerCoupon ) {
+			const forceShow = new URLSearchParams( window.location.search ).get( 'showCouponRedemption' );
+			const partnerCoupon = this.props.partnerCoupon;
+
+			if ( ! this.props.isSiteConnected || forceShow ) {
+				return (
+					<ConnectScreen
+						apiNonce={ this.props.apiNonce }
+						registrationNonce={ this.props.registrationNonce }
+						apiRoot={ this.props.apiRoot }
+						images={ [ '/images/connect-right-partner.png' ] }
+						assetBaseUrl={ this.props.pluginBaseUrl }
+						title={ sprintf(
+							/* translators: %s: Jetpack partner name. */
+							__( 'Welcome to Jetpack %s traveler!', 'jetpack' ),
+							partnerCoupon.partner
+						) }
+						buttonLabel={ sprintf(
+							/* translators: %s: Name of a Jetpack product. */
+							__( 'Set up & redeem %s', 'jetpack' ),
+							partnerCoupon.product.title
+						) }
+						redirectUri={ `admin.php?page=jetpack&partnerCoupon=${ partnerCoupon.coupon_code }` }
+						connectionStatus={ this.props.connectionStatus }
+					>
+						<p>
+							{ sprintf(
+								/* translators: %s: Name of a Jetpack product. */
+								__(
+									'Redeem your coupon and get started with %s for free the first year!',
+									'jetpack'
+								),
+								partnerCoupon.product.title
+							) }
+						</p>
+						<ul>
+							{ partnerCoupon.product.features.map( ( feature, key ) => (
+								<li className="jp-recommendations-product-purchased__feature" key={ key }>
+									{ feature }
+								</li>
+							) ) }
+						</ul>
+						{ this.props.connectionStatus.hasConnectedOwner && (
+							<ActionButton
+								label={ __( 'Redeem coupon', 'jetpack' ) }
+								href={ `https://wordpress.com/checkout/${ this.props.siteRawUrl }/${ partnerCoupon.product.slug }?coupon=${ partnerCoupon.coupon_code }` }
```

Comment:
Quick note so I remember it: Maybe we should point this to the plans page like the connection flow would, so the same Calypso middleware logic will be used for both scenarios, so we can introduce upsell in a uniform way in the future

Reply:
Very good point.

Reply:
@rcoll I don't think the `<ActionButton>` correctly redirected before since it's a `<button>` used outside of a `<form>` which means we just click a button and nothing would happen 😄  I've updated the properties to pass a function to `onClick`  instead.

For now; we still link directly to checkout, so we don't share the same Calypso middleware as the connection flow.
The main reason for this is that the middleware is part of the connection flow/path (`/jetpack/connect/plans`) and I don't want to redirect us to the connection flow, outside of connections. That's a can of worms I don't want to open now 😂 
When we - one day - create an upsell step in Calypso, we'll hopefully have enough control through the Jetpack Redirect to support older plugin versions as well by changing this one.